File: MicroInsight/Login/MILoginViewController.m
//
//  MILoginViewController.m
//  MicroInsight
//
//  Created by Jonphy on 2018/11/19.
//  Copyright © 2018 舒雄威. All rights reserved.
//


#import "MILoginViewController.h"

typedef NS_ENUM(NSInteger,MIModuleType) {
    MIModuleTypeLogin,
    MIModuleTypAssign
};

@interface MILoginViewController ()

@property (weak, nonatomic) IBOutlet UILabel *loginTitle;
@property (weak, nonatomic) IBOutlet UIButton *loginBtn;
@property (weak, nonatomic) IBOutlet UIButton *signInBtn;
@property (weak, nonatomic) IBOutlet UITextField *accountTF;
@property (weak, nonatomic) IBOutlet UITextField *passwordTF;
@property (weak, nonatomic) IBOutlet UITextField *passwordCheckTF;
@property (weak, nonatomic) IBOutlet UIButton *bottomBtn;
@property (weak, nonatomic) IBOutlet UIView *checkPswTFSupView;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *contentViewHeight;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *checkPsdTFHeight;
@property (nonatomic, assign) MIModuleType type;

@end

@implementation MILoginViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    _type = MIModuleTypeLogin;
    _checkPswTFSupView.hidden = YES;
    _contentViewHeight.constant = 200;
    _loginTitle.font = [UIFont captionFontWithName:@"custom" size:20];
}

#pragma mark - IB Action
- (IBAction)loginBtnClick:(UIButton *)sender {
    [self refreshUI];
    [_bottomBtn setTitle:@"登录" forState:UIControlStateNormal];
    _type = MIModuleTypeLogin;
    _checkPswTFSupView.hidden = YES;
    _contentViewHeight.constant = 200;
    sender.titleLabel.font = [UIFont boldSystemFontOfSize:18];
    _signInBtn.titleLabel.font = [UIFont systemFontOfSize:14];
}

- (IBAction)assignBtnClick:(UIButton *)sender {
    [self refreshUI];
    [_bottomBtn setTitle:@"注册" forState:UIControlStateNormal];
    _type = MIModuleTypAssign;
    _checkPswTFSupView.hidden = NO;
    _contentViewHeight.constant = 250;
    sender.titleLabel.font = [UIFont boldSystemFontOfSize:18];
    _loginBtn.titleLabel.font = [UIFont systemFontOfSize:14];
}

- (IBAction)XBtnClick:(UIButton *)sender {
    [self.navigationController popViewControllerAnimated:YES];
}

- (IBAction)bottomBtnClick:(UIButton *)sender {
    [[[UIApplication sharedApplication] keyWindow] endEditing:YES];
    
    if (_accountTF.text.length == 0 || _passwordTF.text.length == 0) {
        [self alertText:@"请输入账号或密码"];
        return;
    }

    if (_type == MIModuleTypAssign) {
        NSString *psd = [_passwordTF.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
        if (psd.length < 6) {
            [self alertText:@"密码不得少于6位"];
            return;
        }
    
        if (![_passwordTF.text isEqualToString:_passwordCheckTF.text]) {
            [self alertText:@"输入的确认密码不正确"];
            return;
        }
    }

    if (_type == MIModuleTypeLogin) {
        [MIRequestManager loginWithUsername:_accountTF.text password:_passwordTF.text completed:^(id  _Nonnull jsonData, NSError * _Nonnull error) {
            
            NSInteger code = [jsonData[@"code"] integerValue];
            if (code == 0) {
                NSDictionary *data = jsonData[@"data"];
                NSDictionary *user = data[@"user"];
                [MILocalData setCurrentLoginUsername:user[@"username"]];
                [MILocalData setCurrentRequestToken:user[@"token"]];
                
                [self.navigationController popViewControllerAnimated:YES];
            } else {
                [self alertText:@"登录失败"];
            }
        }];
    } else {
        [MIRequestManager registerWithUsername:_accountTF.text password:_passwordTF.text completed:^(id  _Nonnull jsonData, NSError * _Nonnull error) {
            
            NSInteger code = [jsonData[@"code"] integerValue];
            if (code == 0) {
                [self loginBtnClick:_loginBtn];
            } else {
                [self alertText:@"注册失败"];
            }
        }];
    }
}

- (IBAction)textfieldEditingChanged:(UITextField *)sender {
    
}

- (void)refreshUI {
    _accountTF.text = nil;
    _passwordTF.text = nil;
    _passwordCheckTF.text = nil;
}

#pragma mark AlertView
- (void)alertText :(NSString *)text{
    
    UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"温馨提示" message:text delegate:self cancelButtonTitle:@"我知道了" otherButtonTitles:nil, nil];
    alert.layer.cornerRadius = 8.0;
    alert.layer.masksToBounds = YES;
    [alert show];
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [[[UIApplication sharedApplication] keyWindow] endEditing:YES];
}

@end

File: MicroInsight/Login/MILoginViewController.swift
//
//  MILoginViewController.swift
//  MicroInsight
//
//  Created by 舒雄威 on 2019/5/12.
//  Copyright © 2019 舒雄威. All rights reserved.
//

import UIKit

class MILoginViewController: MIBaseViewController {
    
    @IBOutlet weak var phoneLab: UILabel!
    @IBOutlet weak var passwordLab: UILabel!
    @IBOutlet weak var phoneTF: UITextField!
    @IBOutlet weak var passwordTF: UITextField!
    @IBOutlet weak var loginBtn: UIButton!
    @IBOutlet weak var authCodeLoginBtn: UIButton!
    @IBOutlet weak var topConstraint: NSLayoutConstraint!
    @IBOutlet weak var otherLoginLab: UILabel!
    @IBOutlet weak var forwardPsdBtn: UIButton!
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        
        configLoginUI()
    }
    
    private func configLoginUI() {
        super.configLeftBarButtonItem(nil)
        super.configRightBarButtonItem(with: .custom, frame: CGRect(x: 0, y: 0, width: 60, height: 30), normalTitle: MILocalData.appLanguage("login_key_2"), normalTitleColor: MIRgbaColor(rgbValue: 0x333333, alpha: 1), highlightedTitleColor: nil, selectedColor: nil, titleFont: 12, normalImage: nil, highlightedImage: nil, selectedImage: nil, touchUpInSideTarget: self, action: #selector(clickRegisterBtn(_:)))
        
        topConstraint.constant = 178 - MINavigationBarHeight(vc: self)
        loginBtn.layer.cornerRadius = 2
        loginBtn.layer.masksToBounds = true
        loginBtn.setTitle(MILocalData.appLanguage("login_key_1"), for: .normal)
        loginBtn.setButtonCustomBackgroudImage(btn: loginBtn, fromColor: MIRgbaColor(rgbValue: 0x72B3E3, alpha: 1), toColor: MIRgbaColor(rgbValue: 0x6DD1CC, alpha: 1))
        authCodeLoginBtn.layer.cornerRadius = 2
        authCodeLoginBtn.layer.masksToBounds = true
        authCodeLoginBtn.setTitle(MILocalData.appLanguage("login_key_6"), for: .normal)
        authCodeLoginBtn.setButtonCustomBackgroudImage(btn: authCodeLoginBtn, fromColor: MIRgbaColor(rgbValue: 0x72B3E3, alpha: 1), toColor: MIRgbaColor(rgbValue: 0x6DD1CC, alpha: 1))
        phoneLab.text = MILocalData.appLanguage("login_key_3")
        passwordLab.text = MILocalData.appLanguage("login_key_4")
        forwardPsdBtn.setTitle(MILocalData.appLanguage("login_key_5"), for: .normal)
        otherLoginLab.text = MILocalData.appLanguage("login_key_7")
    }

    //MARK:按钮点击事件
    //忘记密码
    @IBAction func forwardPassword(_ sender: UIButton) {
        let authCodeVC = MIForwardPasswordViewController.init(nibName: "MIForwardPasswordViewController", bundle: nil)
        self.navigationController?.pushViewController(authCodeVC, animated: true)
    }
    
    //登录
    @IBAction func login(_ sender: UIButton) {
        if !MIHelpTool.validateContactNumber(phoneTF.text) {
            MIHudView.showMsg("请输入正确的手机号码")
            return
        }
        if passwordTF.text!.count == 0 {
            MIHudView.showMsg("请输入密码")
            return
        }
        
        weak var weakSelf = self
        MIRequestManager.login(withUsername: phoneTF.text!, password: passwordTF.text!) { (jsonData, error) in
            
            let dic: [String : AnyObject] = jsonData as! Dictionary
            let code = dic["code"]?.int64Value
            if code == 0 {
                let data = dic["data"]
                let user = data?["user"]
                let model = MIUserInfoModel.yy_model(with: user as! [AnyHashable : Any])
                MILocalData.saveCurrentLoginUserInfo(model)
                
                weakSelf?.navigationController?.popViewController(animated: true)
            } else {
                let msg = dic["message"] as! String
                MIHudView.showMsg(msg)
            }
        }
    }
    
    //验证码登录
    @IBAction func loginByAuthCode(_ sender: UIButton) {
        let authCodeVC = MIAuthCodeLoginViewController.init(nibName: "MIAuthCodeLoginViewController", bundle: nil)
        self.navigationController?.pushViewController(authCodeVC, animated: true)
    }
    
    //qq登录
    @IBAction func loginByQQ(_ sender: UIButton) {
        
        weak var weakSelf = self
        MIThirdPartyLoginManager.share()?.getUserInfo(withWTLoginType: .tencent, result: { (result: Dictionary?, error: String?) in
            
            if (MIHelpTool.isBlankString(error)) {
                MBProgressHUD.showStatus("QQ登录中，请稍后...")
                MIRequestManager.loginByQQ(withOpenId: result?["openId"] as! String, accessToken: result?["accessToken"] as! String, completed: { (jsonData: Any?, err: Error?) in
                    
                    DispatchQueue.main.async {
                        MBProgressHUD.hide()
                    }
                    
                    let dic: [String : AnyObject] = jsonData as! Dictionary
                    let code = dic["code"]?.int64Value
                    if code == 0 {
                        let data = dic["data"]
                        let user = data?["user"]
                        let model = MIUserInfoModel.yy_model(with: user as! [AnyHashable : Any])
                        MILocalData.saveCurrentLoginUserInfo(model)
                        
                        weakSelf?.navigationController?.popToRootViewController(animated: true)
                    } else {
                        let msg = dic["message"] as! String
                        MIHudView.showMsg(msg)
                    }
                })
            } else {
                MIHudView.showMsg(error)
            }
        }, show: self)
    }
    
    //facebook登录
    @IBAction func loginByFacebook(_ sender: UIButton) {
        weak var weakSelf = self
        MIThirdPartyLoginManager.share()?.getUserInfo(withWTLoginType: .facebook, result: { (result: Dictionary?, error: String?) in
            
            if (MIHelpTool.isBlankString(error)) {
                MBProgressHUD.showStatus("Facebook登录中，请稍后...")
                MIRequestManager.loginByFacebook(withDic: result as! Dictionary, completed: { (jsonData: Any?, err: Error?) in
                    
                    DispatchQueue.main.async {
                        MBProgressHUD.hide()
                    }
                    
                    let dic: [String : AnyObject] = jsonData as! Dictionary
                    let code = dic["code"]?.int64Value
                    if code == 0 {
                        let data = dic["data"]
                        let user = data?["user"]
                        let model = MIUserInfoModel.yy_model(with: user as! [AnyHashable : Any])
                        MILocalData.saveCurrentLoginUserInfo(model)
                        
                        weakSelf?.navigationController?.popToRootViewController(animated: true)
                    } else {
                        let msg = dic["message"] as! String
                        MIHudView.showMsg(msg)
                    }
                })
            } else {
                MIHudView.showMsg(error)
            }
        }, show: self)
    }
    
    //微信登录
    @IBAction func loginByWX(_ sender: UIButton) {
        weak var weakSelf = self
        MIThirdPartyLoginManager.share()?.getUserInfo(withWTLoginType: .weiXin, result: { (result: Dictionary?, error: String?) in
            
            if (MIHelpTool.isBlankString(error)) {
                MBProgressHUD.showStatus("微信登录中，请稍后...")
                MIRequestManager.loginByWX(withCode: result?["code"] as! String, completed: { (jsonData: Any?, err: Error?) in
                    
                    DispatchQueue.main.async {
                        MBProgressHUD.hide()
                    }
                    
                    let dic: [String : AnyObject] = jsonData as! Dictionary
                    let code = dic["code"]?.int64Value
                    if code == 0 {
                        let data = dic["data"]
                        let user = data?["user"]
                        let model = MIUserInfoModel.yy_model(with: user as! [AnyHashable : Any])
                        MILocalData.saveCurrentLoginUserInfo(model)

                        weakSelf?.navigationController?.popToRootViewController(animated: true)
                    } else {
                        let msg = dic["message"] as! String
                        MIHudView.showMsg(msg)
                    }
                })
            } else {
                MIHudView.showMsg(error)
            }
        }, show: self)
    }
    
    //注册
    @objc private func clickRegisterBtn(_ sender: UIButton) {
        let authCodeVC = MIRegisterViewController.init(nibName: "MIRegisterViewController", bundle: nil)
        self.navigationController?.pushViewController(authCodeVC, animated: true)
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        self.view.endEditing(true)
    }
}
